src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs

```diff
@@ -7408,4 +7543,282 @@ static class C
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source, targetFramework: TargetFramework.Net90).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Object>", DisplayInvokedMethodTypeArguments(comp));
+    }
+
+    [Fact]
+    public void TypeInference_ReadOnlySpan_ReadOnlySpan_03()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(ReadOnlySpan<object> a, string b) => M1(a, b);
+                void M1<T>(ReadOnlySpan<T> x, T y) { }
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Object>", DisplayInvokedMethodTypeArguments(comp));
+    }
+
+    [Fact]
+    public void TypeInference_ReadOnlySpan_ReadOnlySpan_04()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(ReadOnlySpan<string[]> a, object b) => M1(a, b);
+                void M1<T>(ReadOnlySpan<T[]> x, T y) { }
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source, targetFramework: TargetFramework.Net90).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Object>", DisplayInvokedMethodTypeArguments(comp));
+    }
+
+    [Fact]
+    public void TypeInference_ReadOnlySpan_ReadOnlySpan_05()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(out ReadOnlySpan<string[]> a, object b) => M1(out a, b);
+                void M1<T>(out ReadOnlySpan<T[]> x, T y) => throw null;
+            }
+            """;
+        CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics(
+            // (4,55): error CS0411: The type arguments for method 'C.M1<T>(out ReadOnlySpan<T[]>, T)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+            //     void M(out ReadOnlySpan<string[]> a, object b) => M1(out a, b);
+            Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M1").WithArguments("C.M1<T>(out System.ReadOnlySpan<T[]>, T)").WithLocation(4, 55));
+    }
+
+    [Fact]
+    public void TypeInference_ReadOnlySpan_Span()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(ReadOnlySpan<int> a) => M1(a);
+                void M1<T>(Span<T> x) { }
+            }
+            """;
+        CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics(
+            // (4,36): error CS0411: The type arguments for method 'C.M1<T>(Span<T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+            //     void M(ReadOnlySpan<int> a) => M1(a);
+            Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M1").WithArguments("C.M1<T>(System.Span<T>)").WithLocation(4, 36));
+    }
```

Comment:
>} [](http://example.com/codeflow?start=4&length=1)

Consider testing variance when the array and span are the type arguments. For instance, with the following, for the calls with `I<object[]>` and `IOut<object[]>` I'd expect type inference to succeed but the subsequent argument conversion to fail, and for the call with `IIn<object[]>`, I'd expect type inference to fail.
```csharp
interface I<T> where T : allows ref struct { }
class A
{
    static void M(I<object[]> a) => M1(a);
    static void M1<T>(I<ReadOnlySpan<T>> x) { }
}
interface IOut<out T> where T : allows ref struct { }
class B
{
    static void M(IOut<object[]> a) => M1(a);
    static void M1<T>(IOut<ReadOnlySpan<T>> x) { }
}
interface IIn<in T> where T : allows ref struct { }
class C
{
    static void M(IIn<object[]> a) => M1(a);
    static void M1<T>(IIn<ReadOnlySpan<T>> x) { }
}
```